content/doc/developer/handling-requests/json.adoc

```diff
@@ -0,0 +1,244 @@
+7. `@GET` indicates the HTTP method that is expected, and `@WebMethod` indicates that it is accepting an HTTP request. By default, the JAVA name of the WebMethod is used, but a different name can be specified by using `name =`
+8. `JsonHttpResponse` is a sub class of `HttpResponse` that indicates that the response content will be JSON
+9. An HTTP status can be set in the response.
+10. The method `get-error500` is added in the example to show how to set a error response as JSON.
+
+
+=== Testing with CURL
+
+A `mvn hpi:run` in the plugin should be enough to run it locally.  Assuming that Jenkins is up at http://localhost:8080/jenkins/ , you should have at this point:
+
+[source,bash]
+----
+curl -XGET \
+    -w "\n STATUS:%{http_code}"  \
+    http://localhost:8080/jenkins/custom-api/get-example-param?paramValue=hello
+
+{"message":"I am Jenkins hello"}
+ STATUS:200
+----
+
+=== Example of test with JenkinsRule
+
+[source,java]
+----
+public class JsonAPITest {
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    private static String GET_API_URL = "custom-api/get-example-param?paramValue=hello";
+
+    @Test
+    public void testGetJSON() throws Exception {
+
+        // Testing a simple GET that should answer 200 OK and a json
+        JenkinsRule.JSONWebResponse response = j.getJSON(GET_API_URL);
+        assertTrue(response.getContentAsString().contains("I am JenkinsRule hello"));
+        assertEquals(response.getStatusCode(), 200);
+    }
+
+    @Test
+    public void testAdvancedGetJSON() throws Exception {
+        //Testing a GET that requires the user to be authenticated
+        User admin = User.getById("admin", true);
+        MockAuthorizationStrategy auth = new MockAuthorizationStrategy()
+                .grant(Jenkins.ADMINISTER).everywhere().to(admin);
+
+        j.jenkins.setSecurityRealm(j.createDummySecurityRealm());
+        j.jenkins.setAuthorizationStrategy(auth);
+
+        //We need to setup the webclient
+        //By default if the status code is not ok, WebClient throw an exception
+        //Since we want to assert the error status code, we need to set to false.
+        JenkinsRule.WebClient webClient = j.createWebClient();
+        webClient.setThrowExceptionOnFailingStatusCode(false);
+
+        // - simple call without authentication should be forbidden
+        response = j.getJSON(GET_API_URL, webClient);
+        assertEquals(response.getStatusCode(), 403);
+
+        // - same call but authenticated using withBasicApiToken() should be fine
+        response = j.getJSON(GET_API_URL, webClient.withBasicApiToken(admin));
+        assertEquals(response.getStatusCode(), 200);
+    }
+
+----
+
+=== Handling POST with Jenkins
+
+Starting from the class `JsonAPI` provided for GET example, add:
+
+[source,java]
+----
+@POST
+@WebMethod(name = "")
+public JsonHttpResponse create(@JsonBody JSONObject body) {
+    MyJsonObject parsedBody = (MyJsonObject) body.toBean(MyJsonObject.class);
```

Comment:
💡 This part could be simplified: https://github.com/jenkinsci/jenkins/blob/632033d6ada7c04bbc77ae710dc0a4635fdf9baa/core/src/main/java/jenkins/security/apitoken/LegacyApiTokenAdministrativeMonitor.java#L163

Reply:
I don't see it simple, I see it as different. In my example you have a POJO, when in you link I see it's taking the json Object as an Map of key/value. Both seems ok to me it depends on what is more easy for use on the implementation. I will try to put both of them.

Reply:
ah sorry, I misunderstood, I get it now, thanks !